### DISCUSSION TITLE
"Could not open net0: Not enough Space" on Fujitsu U749

### DISCUSSION THREAD

**Discussion by schorschii** [ipxe/ipxe#327] [Q&A]
Hi, I'm using iPXE via ISC-DHCP-Server for booting Windows and Linux setups in EFI mode. It works fine on several Fujitsu client devices (including Q an E series) but not on Fujitsu U749. Interestingly, it works fine again with the next generation U7410.
Can someone give me a hint what could be the problem? BIOS is already updated to the newest version 2.21.
![IMG_20210420_110142_converted](https://user-images.githubusercontent.com/17480795/115378885-5159ab80-a1d1-11eb-94d5-4d717662280d.jpg)

And this is the output when compiling iPXE with DEBUG=intel:
![IMG_20210420_110642_converted](https://user-images.githubusercontent.com/17480795/115378913-574f8c80-a1d1-11eb-90ec-fe0bc4e1b60a.jpg)

Thanks!

**Comment by NiKiZe** [comment]
Full message: "Not enough Space (http://ipxe.org/31654089)"

**Reply by NiKiZe** [reply]
A game of follow the rabbit
https://github.com/ipxe/ipxe/blob/1192edf394aaebde666c6aafda1c8b782b08a845/src/core/malloc.c#L313
https://github.com/ipxe/ipxe/blob/1192edf394aaebde666c6aafda1c8b782b08a845/src/core/malloc.c#L624

please try with DEBUG=malloc:3

And maybe also try to run memtest or similar (this smells a bit like hardware issues, if not it would be great to be able to rule that one out)

**Reply by mcb30** [reply]
So, there's no way that the fixed-size internal heap could possibly be exhausted so early on.  This rules out the possibility that it's `alloc_etherdev()` that's failing.  The failure must therefore be from `dma_alloc()`.

For a PCI device under UEFI, the implementation of `dma_malloc()` will be provided by `efipci_dma_malloc()` in `efi_pci.c`.  @schorschii could you try building with `DEBUG=intel,efi_pci`?

**Reply by mcb30** [reply]
I'm guessing the issue might be that the call to `AllocateBuffer()` allocates a number of pages (based on rounding up the required length), but the call to `Map()` specifies the exact required length.  It's plausible that the UEFI firmware's implementation of `Map()` is complaining because the (byte-aligned) map length doesn't exactly match the (page-aligned) allocation length.  If so, then it would be a fairly simple change.

**Reply by schorschii** [reply]
Thanks for your fast responses.
@NiKiZe I tried it with a second U749 and ist shows the same error. PassMark MemTest86 shows no errors. Starting iPXE with `DEBUG=malloc:3` displays very much debug info, on what should I pay attention to?

@mcb30 `DEBUG=intel,efi_pci` brings up the following output:
![IMG_20210420_132116](https://user-images.githubusercontent.com/17480795/115388772-8d463e00-a1dc-11eb-8975-dee3fb38e5f8.jpg)

"cannot map 5a962000+100: Error 0x7f10e082"

**Reply by mcb30** [reply]
> @mcb30 `DEBUG=intel,efi_pci` brings up the following output:
> "cannot map 5a962000+100: Error 0x7f10e082"

Thanks.  That's an `EFI_INVALID_PARAMETER` returned from the `Map()` call, which would fit with my hypothesis.  Could you try with this patch:

```diff
diff --git a/src/interface/efi/efi_pci.c b/src/interface/efi/efi_pci.c
index 4adee0fd8..ef02c9c0e 100644
--- a/src/interface/efi/efi_pci.c
+++ b/src/interface/efi/efi_pci.c
@@ -523,7 +523,8 @@ static void * efipci_dma_alloc ( struct dma_device *dma,
 
        /* Map buffer */
        if ( ( rc = efipci_dma_map ( dma, map, virt_to_phys ( addr ),
-                                    len, DMA_BI ) ) != 0 )
+                                    ( pages * EFI_PAGE_SIZE ),
+                                    DMA_BI ) ) != 0 )
                goto err_map;
 
        /* Increment allocation count (for debugging) */
```

**Reply by schorschii** [reply]
Unfortunately, I get almost the same error with the patch: `cannot map 5a962000+1000: Error 0x7f10e082`
Only `+100` changed to `+1000`

**Reply by mcb30** [reply]
> Unfortunately, I get almost the same error with the patch: `cannot map 5a962000+1000: Error 0x7f10e082`
> Only `+100` changed to `+1000`

That's a shame.  I've merged this change anyway as commit https://github.com/ipxe/ipxe/commit/3efdbef2f since it won't hurt and may avoid future problems on other platforms.

The only other possibility I can think of is that the `Map()` implementation is examining some of the output parameters and (erroneously) checking that they are initialised to zero.  You could try with this patch (on current master):

```diff
diff --git a/src/interface/efi/efi_pci.c b/src/interface/efi/efi_pci.c
index ef02c9c0e..19246633d 100644
--- a/src/interface/efi/efi_pci.c
+++ b/src/interface/efi/efi_pci.c
@@ -418,6 +418,8 @@ static int efipci_dma_map ( struct dma_device *dma, struct dma_mapping *map,
 
        /* Map buffer (if non-zero length) */
        count = len;
+       bus = 0;
+       mapping = NULL;
        if ( len ) {
                if ( ( efirc = pci_io->Map ( pci_io, op, phys_to_virt ( addr ),
                                             &count, &bus, &mapping ) ) != 0 ) {
```

If that doesn't work, then we'll probably need to use the `imgmem` command to extract a copy of the PCI bus driver for disassembly.

**Reply by schorschii** [reply]
Thanks for your help. Unfortunately, also this patch did not eliminate the error.

Next, I tested it without error handling, which means I commented out `goto err_map;` on line 428... and IT WORKS! iPXE boots my Linux and Windows systems on the U749 devices.

I'm sure this is not a good idea for a production environment, I only want to share my observation.
How can I use the `imgmem` command to provide more information?

**Reply by mcb30** [reply]
> How can I use the `imgmem` command to provide more information?

Actually, it's probably easier to use the UEFI shell directly.  Boot into a full version of the UEFI shell (e.g. using `chain http://boot.ipxe.org/Shell.efi`), and then:
* use the `drivers` command to locate the PCI bus driver (most likely called `PciBusDxe`)
* use the `dh -v` command on the driver handle to identify `ImageBase` and `ImageSize`
* use the `dmem` command and shell redirection to dump the PCI bus driver loaded image to a USB key

Also add some debug in iPXE to print out the value of the `pci_io->Map` pointer: check that this function pointer is within the range of `ImageBase`+`ImageSize` and send it to me along with the dumped PCI bus driver.

**Reply by schorschii** [reply]
OK, I managed to dump the PCI bus driver:

```
Shell> drivers
            T   D
D           Y C I
R           P F A
V  VERSION  E G G #D #C DRIVER NAME                         IMAGE NAME
== ======== = = = == == =================================== ==========
...
211 0000000A B - -  1 26 PCI Bus Driver                      PciBusDxe
...

Shell> dh -v 211
211: 5B5B4B18
ComponentName2(6A7A2D50)
ComponentName(6A7A2D38)
DriverBinding(6A7A2CD0)
ImageDevicePath(5B5B4718)
  MemoryMapped(0xB,0x50BAF000,0x51B4DFFF)/FvFile(93B80004-9FB3-11D4-9A3A-0090273FC14D)
LoadedImage(5B5B4840)
  Name..........: PciBusDxe
  Revision......: 0x00001000
  ParentHandle..: 5CCDEF98
  SystemTable...: 6C8EE018
  DeviceHandle..: 5CCAC818
  FilePath......: FvFile(93B80004-9FB3-11D4-9A3A-0090273FC14D)
  PdbFileName...: <null string>
  OptionsSize...: 0
  LoadOptions...: 0
  ImageBase.....: 6A792000
  ImageSize.....: 12700
  CodeType......: EfiBootServicesCode
  DataType......: EfiBootServicesData
  Unload........: 0

Shell> dmem 6A792000 12700 > dmem.txt
```
Dump File: [dmem.txt](https://github.com/ipxe/ipxe/files/6350288/dmem.txt)

> Also add some debug in iPXE to print out the value of the pci_io->Map pointer: check that this function pointer is within the range of ImageBase+ImageSize

For this, I need some more help. How do I do that exactly?

**Reply by mcb30** [reply]
> OK, I managed to dump the PCI bus driver:
> Dump File: [dmem.txt](https://github.com/ipxe/ipxe/files/6350288/dmem.txt)

Perfect; thank you!

> > Also add some debug in iPXE to print out the value of the pci_io->Map pointer: check that this function pointer is within the range of ImageBase+ImageSize
> 
> For this, I need some more help. How do I do that exactly?

Sorry; I should have been more explicit.  Try something like this patch, and build with `DEBUG=efi_pci`:

```diff
diff --git a/src/interface/efi/efi_pci.c b/src/interface/efi/efi_pci.c
index ef02c9c0e..db32dc3e6 100644
--- a/src/interface/efi/efi_pci.c
+++ b/src/interface/efi/efi_pci.c
@@ -425,6 +425,7 @@ static int efipci_dma_map ( struct dma_device *dma, struct dma_mapping *map,
 			DBGC ( pci, "EFIPCI " PCI_FMT " cannot map %08lx+%zx: "
 			       "%s\n", PCI_ARGS ( pci ), addr, len,
 			       strerror ( rc ) );
+			DBGC ( pci, "*** Map = %p\n", pci_io->Map );
 			goto err_map;
 		}
 	} else {
```

**Reply by schorschii** [reply]
OK, this gives me:
```
EFIPCI 0000:00:1f.6 cannot map 5a96f000+1000: Error 0x7f10e082
*** Map = 0x6a79cfb8
Could not open net0: Not enough space
```
So the function pointer seems to be within the range.

**Reply by mcb30** [reply]
Thanks.  From looking at the disassembly, it seems to be just the vanilla EDK2 PciBusDxe driver.  There is nothing within the `pci_io->Map` function itself that could be raising the `EFI_INVALID_PARAMETER`, which means that it must be coming from either the underlying `EFI_PCI_ROOT_BRIDGE_IO_PROTOCOL` or the (vendor-specific) IOMMU protocol.  The most likely suspect is the vendor-specific IOMMU protocol, since vendor UEFI code is often of very dubious quality.

Are there any IOMMU-related options within the UEFI setup menu?  (May be labelled something like "DMAr" or "memory protection")  If so, do they make any difference to the error?

**Reply by schorschii** [reply]
Looking at the UEFI settings, I couldn't find any option like "IOMMU", "DMAr" or "memory protection" on this model. But I tried to change some other settings and there was a surprise: disabling the "Intel VT-d" (below "Virtualization Technology") lets iPXE start the network interface. This is strange and I think this is a model specific UEFI issue too, because I don't need to disable it on models U748 and U7410.

I don't know if it's reasonable to implement a patch in iPXE just for this (buggy) U749 model. My workflow is now to disable "VT-d" for the installation and then enable it again after installation finished.

Thanks a lot for your support!

**Reply by mcb30** [reply]
"VT-d" is another term for the IOMMU so yes, that makes sense.

It might be possible to figure out precisely why the U749 is rejecting the DMA mapping request, but I'd need remote access to the hardware in question to carry out the debugging.

**Reply by schorschii** [reply]
OK, how do we do that? I install a Linux system and you access via TeamViewer?

**Reply by mcb30** [reply]
You'd need to be able to reproduce the issue in a system that has some kind of lights-out remote access (e.g. HP iLO, Dell DRAC, or an external IP-KVM box), and then provide access to that along with ssh access to a Linux system on the same network.  Is this something you might be able to provide?

**Comment by srett** [comment]
One of our users reported a very similar issue in their setup. They get http://ipxe.org/31cb4089 instead (cb instead of 65, because intelxl), since it happens with a PCIe Intel X520-DA2 in a Fujitsu P910. We've got a P910 in our lab, but our X520 is M.I.A., but hopefully we'll get it back next week and will be able to reproduce this.

Interesting insights:
 * ipxe.efi boots fine if using the onboard e1000e instead (according to user, confirmed with our model)
 * I tried a Broadcom NetXtreme II BCM57810 just for fun, and the BIOS would freeze during POST. Works fine in another EFI machine.
 * It worked fine with "some iPXE revision from last August" (according to user)

What immediately came up when skimming the git history is the whole DMA-API restructuring that was introduced starting in November. I can do a git bisect if that helps, but it'll probably just bring up 03314e8da or 24ef743778 which wouldn't help that much...
Will report back once I got the card.

**Reply by mcb30** [reply]
Thanks.  Yes, it's expected that this failure would be hidden prior to the addition of IOMMU support to iPXE, since older versions of iPXE will just assume that the whole of memory is identity-mapped and are therefore "immune" to buggy UEFI implementations that return spurious IOMMU errors.

It's interesting that it would work with the onboard card, though.  I'd expect it to either work on all PCI devices or fail on all PCI devices.  But this is just based on an expectation of consistency from the UEFI firmware implementation, and that may not be a justified expectation.  :slightly_smiling_face:

**Reply by srett** [reply]
It took a very long time, but after some back and forth, and a correction by the user that they are actually running an X710 and not an X520, we ordered such a card and tried everything to recreate that problem, without success so far.
We requested dmidecode, lspci and lshw output from that machine and matched our BIOS version and memory configuration, and still were unable to reproduce. I guess the only thing that might make a difference now is BIOS settings, but which ones is anyone's guess.

**Reply by mcb30** [reply]
> It took a very long time, but after some back and forth, and a correction by the user that they are actually running an X710 and not an X520, we ordered such a card and tried everything to recreate that problem, without success so far. We requested dmidecode, lspci and lshw output from that machine and matched our BIOS version and memory configuration, and still were unable to reproduce. I guess the only thing that might make a difference now is BIOS settings, but which ones is anyone's guess.

Do you definitely have the IOMMU enabled in the BIOS settings?  (May be labelled "VT-d", "DMAr", "DMA memory protection", "Device protection", or any of quite a few other synonyms.)

**Reply by srett** [reply]
Yes it's enabled, the machine is usually running VMs with PCI passthough. But just for fun I just *disabled* VT-d – no change, still working fine. Unfortunately the machine is too old to have a "export BIOS settings to USB" option, otherwise I could ask the user to export their settings and send them over. Doing it manually is probably impractical.

( Also, they're currently just running an old iPXE build, so from their POV, the problem is fixed. :-/ )
